modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningMetadataExtractor.java

```diff
@@ -0,0 +1,560 @@
+                PruningColumnSets columnSets = (PruningColumnSets) res;
+                candidates.addAll(columnSets.candidates);
+            }
+        }
+
+        /** Combines this column sets with the given result. Argument can be either PruningColumnSet or PruningColumnSets. */
+        void combine(Result res) {
+            // Combine (AND) merges each existing column sets with each candidate column set an for each of those
+            // produces new result
+
+            // lhs    :  [c1 = 1, c2 = 2]
+            // rhs    :  [c3 = 3]
+            // result :  [c1 = 1, c2 = 2, c3 = 3]
+
+            // lhs    :  [c1 = 1, c2 = 2]
+            // rhs    :  [c3 = 3], [c4 = 4]
+            // result :  [c1 = 1, c2 = 2, c3 = 3], [c1 = 1, c2 = 2, c4 = 4]
+
+            // ADD also does not allow to have the same columns with different values:
+
+            // lhs    :  [c1 = 1, c2 = 2]
+            // rhs    :  [c1 = 2]
+            // result :  conflict -> c1 # c1 can't be equal to both 1 and 2
+
+            // lhs    :  [c1 = 1, c2 = 2]
+            // rhs    :  [c1 = 2, c2 = 3]
+            // result :  conflict -> c1, c2
+
+            if (candidates.isEmpty()) {
+                if (res instanceof PruningColumnSet) {
+                    PruningColumnSet columnSet = (PruningColumnSet) res;
+                    candidates.add(columnSet);
+                } else {
+                    PruningColumnSets other = (PruningColumnSets) res;
+                    candidates.addAll(other.candidates);
+                }
+            } else {
+
+                if (conflict) {
+                    return;
+                }
+
+                PruningColumnSets other;
+                if (res instanceof PruningColumnSet) {
+                    other = new PruningColumnSets();
+                    other.candidates.add((PruningColumnSet) res);
+                } else {
+                    other = (PruningColumnSets) res;
+                }
+
+                List<PruningColumnSet> newOutput = new ArrayList<>();
+
+                for (PruningColumnSet candidate : other.candidates) {
+                    for (PruningColumnSet val : candidates) {
+                        for (Entry<RexNode> ckv : candidate.columns.int2ObjectEntrySet()) {
+                            if (!val.columns.containsKey(ckv.getIntKey())) {
+                                continue;
+                            }
+
+                            Object existing = val.columns.get(ckv.getIntKey());
+                            if (!Objects.equals(existing, ckv.getValue())) {
+                                conflict = true;
+                                return;
+                            }
+                        }
+
+                        Int2ObjectMap<RexNode> newValue = new Int2ObjectArrayMap<>(val.columns);
+                        newValue.putAll(candidate.columns);
+
+                        PruningColumnSet newColSet = new PruningColumnSet(newValue);
```

Comment:
```suggestion
                        Int2ObjectMap<RexNode> newValue = new Int2ObjectArrayMap<>(val.columns.size() + candidate.columns.size());
                        newValue.putAll(val.columns);
                        newValue.putAll(candidate.columns);

                        PruningColumnSet newColSet = new PruningColumnSet(newValue);
```
and maybe move to a separate method `concat(PruningColumnSet, PruningColumnSet)`